play-json-circe/src/main/scala/com/evolutiongaming/util/PlayCirceAstConversions.scala

```diff
@@ -0,0 +1,41 @@
+package com.evolutiongaming.util
+
+import io.circe.{Json => CirceJson}
+import play.api.libs.{json => PlayJson}
+
+object PlayCirceAstConversions {
+  def circeToPlay(circeJson: CirceJson): PlayJson.JsValue =
+    circeJson.fold(
+      jsonNull = PlayJson.JsNull,
+      jsonBoolean = b => PlayJson.JsBoolean(b),
+      jsonNumber = n => n.toBigDecimal.map(PlayJson.JsNumber).getOrElse(PlayJson.JsNumber(n.toDouble)),
+      jsonString = s => PlayJson.JsString(s),
+      jsonArray = as => PlayJson.JsArray(as.map(circeToPlay)),
+      jsonObject = o => PlayJson.JsObject(o.toIterable.map { case (k, v) => (k, circeToPlay(v)) }.toSeq),
+    )
+
+  def playToCirce(value: PlayJson.JsValue): CirceJson =
```

Comment:
tailrec?

Reply:
`circeToPlay` is not in tail position

Reply:
this is a question - we either decide that we are not going to support ~1000 nested elements or try to rewrite so it could be tailrec-ed

Reply:
Well, I don't know.  

It's another part which was not changed compared to original solution and worked so far without blowing the stack. 

If we keep it, then both these conversions deserve disclaimer about safety in scaladoc.

Reply:
Suggestion:

```
  @tailrec
  private def playToCirceSeq(seq: scala.collection.IndexedSeq[PlayJson.JsValue], acc: List[CirceJson] = Nil): Iterable[CirceJson] = {
    seq.headOption match {
      case Some(value) => 
        playToCirceSeq(seq.tail, playToCirce(value) :: acc)
      case None        => 
        acc.reverse
    }
  }

  @tailrec
  private def playToCirceKeyValue(list: List[(String, JsValue)], acc: List[(String, CirceJson)] = Nil): Iterable[(String, CirceJson)] = {
    list match {
      case head :: tail =>
        val (k, v) = head
        playToCirceKeyValue(tail, (k, playToCirce(v)) :: acc)
      case Nil          =>
        acc.reverse
    }
  }

  def playToCirce(value: PlayJson.JsValue): CirceJson = {
    value match {
      case PlayJson.JsNull =>
        CirceJson.Null

      case PlayJson.JsTrue =>
        CirceJson.True

      case PlayJson.JsFalse =>
        CirceJson.False

      case PlayJson.JsNumber(value) =>
        CirceJson.fromBigDecimal(value)

      case PlayJson.JsString(value) =>
        CirceJson.fromString(value)

      case PlayJson.JsArray(value) =>
        CirceJson.fromValues(playToCirceSeq(value))

      case PlayJson.JsObject(value) =>
        CirceJson.fromFields(playToCirceKeyValue(value.toList))
    }
  }
```

Data structures used could be changed (e.g. to avoid `List.reverse`).